File: ui/components/chessboard.py
"""Chessboard.js Streamlit component wrapper."""

from pathlib import Path
from typing import Any

import chess
import streamlit as st
import streamlit.components.v1 as components

# Get the component directory
_COMPONENT_DIR = Path(__file__).parent / "chessboard"

# Declare the component
_chessboard_component = components.declare_component(
    "chessboard",
    path=str(_COMPONENT_DIR)
)


def chessboard(
    fen: str = chess.STARTING_FEN,
    orientation: str = "white",
    last_move: str | None = None,
    check_square: str | None = None,
    legal_moves: list[str] | None = None,
    show_coordinates: bool = True,
    draggable: bool = True,
    animate: bool = True,
    on_move: callable | None = None,
    key: str | None = None
) -> dict[str, Any]:
    """
    Render an interactive chessboard using chessboard.js.
    
    Args:
        fen: FEN string for the position
        orientation: "white" or "black" - board orientation
        last_move: UCI move string to highlight (e.g., "e2e4")
        check_square: Square in check (e.g., "e8")
        legal_moves: List of legal move UCI strings to highlight
        show_coordinates: Show rank/file coordinates
        draggable: Allow piece dragging
        animate: Animate piece movements
        on_move: Callback function(move_uci, source, target, piece) -> bool
        key: Unique key for the component
    
    Returns:
        Dict with component state including position, lastMove, etc.
    """

    # Prepare config
    config = {
        "position": fen,
        "orientation": orientation,
        "showCoordinates": show_coordinates,
        "draggable": draggable,
        "animate": animate,
        "pieceTheme": "https://chessboardjs.com/img/chesspieces/wikipedia/{piece}.png",
    }

    # Call the component
    component_value = _chessboard_component(
        config=config,
        lastMove=last_move,
        checkSquare=check_square,
        legalMoves=legal_moves or [],
        clearHighlights=False,
        key=key,
        default={"position": fen, "lastMove": last_move}
    )

    return component_value or {}


def render_chessboard_js():
    """Inject the chessboard.js library and component JS."""
    # This should be called once in the app
    st.components.v1.html("""
    <script src="https://cdnjs.cloudflare.com/ajax/libs/chessboard-js/1.0.0/chessboard-1.0.0.min.js"></script>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/chessboard-js/1.0.0/chessboard-1.0.0.min.css" />
    """, height=0)


if __name__ == "__main__":
    import chess
    import streamlit as st

    st.title("Chessboard.js Demo")

    # Render chessboard.js library
    render_chessboard_js()

    # Demo board
    board = chess.Board()

    col1, col2 = st.columns([2, 1])

    with col1:
        result = chessboard(
            fen=board.fen(),
            orientation="white",
            last_move="e2e4" if board.move_stack else None,
            check_square=board.king(board.turn) if board.is_check() else None,
            legal_moves=[m.uci() for m in board.legal_moves[:5]],
            key="demo_board"
        )
        st.write("Component result:", result)

    with col2:
        st.write("**Controls**")
        if st.button("Make Move e2e4"):
            board.push_uci("e2e4")
            st.rerun()

        if st.button("Flip Board"):
            st.rerun()

        st.write("**Position:**")
        st.code(board.fen())

File: ui/components/chessboard/chessboard.js
/**
 * Chessboard.js Streamlit Component
 * Interactive chess board with drag-and-drop, animations, and coordinates
 */

// Chessboard component for Streamlit
class ChessboardComponent {
    constructor() {
        this.board = null;
        this.game = null;
        this.config = {
            position: 'start',
            orientation: 'white',
            showCoordinates: true,
            draggable: true,
            animations: true,
            pieceTheme: 'https://chessboardjs.com/img/chesspieces/wikipedia/{piece}.png',
        };
        this.onMove = null;
        this.onSnapEnd = null;
    }

    init(container, config = {}) {
        this.config = { ...this.config, ...config };
        
        // Create chessboard.js board
        this.board = Chessboard(container, {
            position: this.config.position,
            orientation: this.config.orientation,
            showCoordinates: this.config.showCoordinates,
            draggable: this.config.draggable,
            pieceTheme: this.config.pieceTheme,
            onDrop: this.handleDrop.bind(this),
            onSnapEnd: this.handleSnapEnd.bind(this),
            onMoveEnd: this.handleMoveEnd.bind(this),
        });

        // Add animation CSS
        this.injectStyles();
    }

    injectStyles() {
        if (document.getElementById('chessboard-animations')) return;
        
        const style = document.createElement('style');
        style.id = 'chessboard-animations';
        style.textContent = `
            .chessboard-piece {
                transition: transform 0.2s ease-out, opacity 0.1s ease-out !important;
            }
            .chessboard-piece.animating {
                transition: transform 0.3s cubic-bezier(0.25, 0.46, 0.45, 0.94) !important;
            }
            .square-55d63.highlight-last-move {
                box-shadow: inset 0 0 0 4px #ffcc00 !important;
            }
            .square-55d63.highlight-check {
                box-shadow: inset 0 0 0 4px #ff3333 !important;
            }
            .square-55d63.highlight-legal {
                background: rgba(0, 200, 0, 0.15) !important;
            }
            .chessboard-squares .square-55d63 {
                transition: background-color 0.1s ease, box-shadow 0.1s ease;
            }
        `;
        document.head.appendChild(style);
    }

    handleDrop(source, target, piece, newPos, oldPos, orientation) {
        // Validate move
        const move = source + target;
        
        // Check if move is legal (basic UCI validation)
        if (!this.isValidUCI(move)) {
            return 'snapback';
        }

        // Call custom move handler
        if (this.onMove) {
            const result = this.onMove(move, source, target, piece);
            if (result === false) {
                return 'snapback';
            }
        }

        return move;
    }

    handleSnapEnd() {
        if (this.onSnapEnd) {
            this.onSnapEnd(this.board.position());
        }
    }

    handleMoveEnd(oldPos, newPos) {
        // Highlight last move
        this.highlightLastMove(oldPos, newPos);
    }

    highlightLastMove(from, to) {
        // Remove previous highlights
        document.querySelectorAll('.square-55d63.highlight-last-move').forEach(el => {
            el.classList.remove('highlight-last-move');
        });
        
        // Add new highlights
        const fromEl = document.querySelector(`.square-${from}`);
        const toEl = document.querySelector(`.square-${to}`);
        
        if (fromEl) fromEl.classList.add('highlight-last-move');
        if (toEl) toEl.classList.add('highlight-last-move');
    }

    highlightCheck(square) {
        const el = document.querySelector(`.square-${square}`);
        if (el) el.classList.add('highlight-check');
    }

    highlightLegalMoves(squares) {
        squares.forEach(sq => {
            const el = document.querySelector(`.square-${sq}`);
            if (el) el.classList.add('highlight-legal');
        });
    }

    clearHighlights() {
        document.querySelectorAll('.square-55d63.highlight-last-move, .square-55d63.highlight-check, .square-55d63.highlight-legal').forEach(el => {
            el.classList.remove('highlight-last-move', 'highlight-check', 'highlight-legal');
        });
    }

    isValidUCI(move) {
        // Basic UCI format: e2e4, e7e8q
        return /^[a-h][1-8][a-h][1-8][qrbn]?$/.test(move);
    }

    setPosition(fen, animate = true) {
        this.board.position(fen, animate);
    }

    getPosition() {
        return this.board.position();
    }

    flip() {
        this.board.flip();
    }

    resize() {
        this.board.resize();
    }

    destroy() {
        this.board.destroy();
    }

    // Animation helpers
    animateMove(from, to, piece) {
        const pieceEl = document.querySelector(`.square-${from} .chessboard-piece`);
        if (!pieceEl) return Promise.resolve();

        const fromSquare = document.querySelector(`.square-${from}`);
        const toSquare = document.querySelector(`.square-${to}`);
        
        if (!fromSquare || !toSquare) return Promise.resolve();

        const fromRect = fromSquare.getBoundingClientRect();
        const toRect = toSquare.getBoundingClientRect();

        const dx = toRect.left - fromRect.left;
        const dy = toRect.top - fromRect.top;

        pieceEl.classList.add('animating');
        pieceEl.style.transform = `translate(${dx}px, ${dy}px)`;

        return new Promise(resolve => {
            setTimeout(() => {
                pieceEl.classList.remove('animating');
                pieceEl.style.transform = '';
                resolve();
            }, 300);
        });
    }
}

// Export for Streamlit
window.ChessboardComponent = ChessboardComponent;

// Streamlit component interface
function createChessboardComponent() {
    return {
        name: 'chessboard',
        render: function(container, props) {
            if (!window.chessboardInstance) {
                window.chessboardInstance = new ChessboardComponent();
                window.chessboardInstance.init(container, props.config);
                
                // Set up callbacks
                window.chessboardInstance.onMove = function(move, source, target, piece) {
                    if (props.onMove) {
                        props.onMove(move, source, target, piece);
                    }
                };
                
                window.chessboardInstance.onSnapEnd = function(position) {
                    if (props.onSnapEnd) {
                        props.onSnapEnd(position);
                    }
                };
            } else {
                // Update config if changed
                if (props.config) {
                    if (props.config.position) {
                        window.chessboardInstance.setPosition(props.config.position, props.config.animate !== false);
                    }
                    if (props.config.orientation) {
                        window.chessboardInstance.board.orientation(props.config.orientation);
                    }
                }
            }

            // Handle programmatic moves
            if (props.lastMove) {
                const [from, to] = [props.lastMove.slice(0, 2), props.lastMove.slice(2, 4)];
                window.chessboardInstance.highlightLastMove(from, to);
            }

            if (props.checkSquare) {
                window.chessboardInstance.highlightCheck(props.checkSquare);
            }

            if (props.legalMoves) {
                window.chessboardInstance.highlightLegalMoves(props.legalMoves);
            }

            if (props.clearHighlights) {
                window.chessboardInstance.clearHighlights();
            }

            return container;
        },
        dispose: function() {
            if (window.chessboardInstance) {
                window.chessboardInstance.destroy();
                window.chessboardInstance = null;
            }
        }
    };
}

// Register with Streamlit if available
if (typeof Streamlit !== 'undefined') {
    Streamlit.components.v1.registerComponent('chessboard', createChessboardComponent());
}